controllers/storageclient_controller.go

```diff
@@ -0,0 +1,307 @@
+func (s *StorageClientReconciler) onboardConsumer(instance *v1alpha1.StorageClient, externalClusterClient *providerClient.OCSProviderClient) (reconcile.Result, error) {
+
+	// TODO Need to find a way to this one as its OCP dependent
+	clusterVersion := &configv1.ClusterVersion{}
+	err := s.Client.Get(s.ctx, types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		s.Log.Error(err, "failed to get the clusterVersion version of the OCP cluster")
+		return reconcile.Result{}, fmt.Errorf("failed to get the clusterVersion version of the OCP cluster: %v", err)
+	}
+
+	name := fmt.Sprintf("storageconsumer-%s", clusterVersion.Spec.ClusterID)
+	// TODO: remove hardcoding of the capacity
+	response, err := externalClusterClient.OnboardConsumer(
+		s.ctx, instance.Spec.OnboardingTicket, name, "1T")
+	if err != nil {
+		if st, ok := status.FromError(err); ok {
+			s.logGrpcErrorAndReportEvent(instance, OnboardConsumer, err, st.Code())
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to onboard consumer: %v", err)
+	}
+
+	if response.StorageConsumerUUID == "" {
+		err = fmt.Errorf("storage provider response is empty")
+		s.Log.Error(err, "empty response")
+		return reconcile.Result{}, err
+	}
+
+	instance.Status.ConsumerID = response.StorageConsumerUUID
+	instance.Status.Phase = v1alpha1.StorageClientOnboarding
+
+	s.Log.Info("onboarding started")
+	return reconcile.Result{Requeue: true}, nil
+}
+
+func (s *StorageClientReconciler) acknowledgeOnboarding(instance *v1alpha1.StorageClient, externalClusterClient *providerClient.OCSProviderClient) (reconcile.Result, error) {
+
+	_, err := externalClusterClient.AcknowledgeOnboarding(s.ctx, instance.Status.ConsumerID)
+	if err != nil {
+		if st, ok := status.FromError(err); ok {
+			s.logGrpcErrorAndReportEvent(instance, AcknowledgeOnboarding, err, st.Code())
+		}
+		s.Log.Error(err, "Failed to acknowledge onboarding.")
+		return reconcile.Result{}, fmt.Errorf("failed to acknowledge onboarding: %v", err)
+	}
+	instance.Status.Phase = v1alpha1.StorageClientConnected
+
+	s.Log.Info("Onboarding is acknowledged successfully.")
+	return reconcile.Result{Requeue: true}, nil
+}
+
+// offboardConsumer makes an API call to the external storage provider cluster for offboarding
+func (s *StorageClientReconciler) offboardConsumer(instance *v1alpha1.StorageClient, externalClusterClient *providerClient.OCSProviderClient) (reconcile.Result, error) {
+
+	_, err := externalClusterClient.OffboardConsumer(s.ctx, instance.Status.ConsumerID)
+	if err != nil {
+		if st, ok := status.FromError(err); ok {
+			s.logGrpcErrorAndReportEvent(instance, OffboardConsumer, err, st.Code())
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to offboard consumer: %v", err)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (s *StorageClientReconciler) logGrpcErrorAndReportEvent(instance *v1alpha1.StorageClient, grpcCallName string, err error, errCode codes.Code) {
```

Comment:
The fact there's an `And` in the function name immediately flags that this should be split. Looking at it deeper, the first part of the function is just a selection of strings, and the actual work of the function is two lines. Similarly, only one of these groups of lines is ever used more than once.

I can kind of understand want to have a single place to abstract error reporting, but I don't think this is really necessary.  Given that this is only even used in three places, that also indicates this is probably needless abstraction. I could still see an argument for making it easy to write and reference our error reporting strings, so it might make sense to just define some struct like this, and define a few of these at the top of some file:

```go
type struct StorageClientGrpcEvent {
    msg string
    eventReason string
    eventType string
    errorCode codes.Code
}
```

Then you could go to the three places this function is currently used and replace it with the following, which avoids us falling into the trap of constraining ourselves to an abstraction format we might not consistently need:
```go
		if st, ok := status.FromError(err); ok {
			var errStruct = SomeVarName
			s.Log.Error(err, "StorageProvider:"+grpcCallName+":"+errStruct.msg)
			s.recorder.ReportIfNotPresent(instance, errStruct.eventType, errStruct.eventReason, errStruct.msg)
		}
		return reconcile.Result{}, fmt.Errorf(errStruct.msg, err)
```

If you agree with this assessment, this can also come in as a future PR from you or someone else. If you don't, that's valid, let's hear why.

Reply:
Yes, let's take it as a follow-up, will open a GitHub issue to track it.

Reply:
Issue #8 has been opened.